pkg/internal/ibmblockcsi/default_setter.go

```diff
@@ -17,32 +17,109 @@
 package ibmblockcsi
 
 import (
+	"strings"
+
+	csiv1 "github.com/IBM/ibm-block-csi-operator/pkg/apis/csi/v1"
 	"github.com/IBM/ibm-block-csi-operator/pkg/config"
+	corev1 "k8s.io/api/core/v1"
 )
 
 // SetDefaults set defaults if omitted in spec, returns true means CR should be updated on cluster.
 // Replace it with kubernetes native default setter when it is available.
 // https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#defaulting
-func (c *IBMBlockCSI) SetDefaults() bool {
+func (c *IBMBlockCSI) SetDefaults(platform string) bool {
 	// repository is mandatory, tag is optional, but if repository is not set
 	// and tag is set, the tag will be overrided to the default one.
-	changed := false
+
+	changed := c.clearPreviousVersion(platform)
 
 	// if controller is empty
 	if c.Spec.Controller.Repository == "" {
-		c.Spec.Controller.Repository = config.ControllerRepository
-		c.Spec.Controller.Tag = config.ControllerTag
+		regAndTag := strings.Split(c.GetDefaultImageByPlatformAndName(platform, config.Controller), ":")
+		c.Spec.Controller.Repository = regAndTag[0]
+		c.Spec.Controller.Tag = regAndTag[1]
 
 		changed = true
 	}
 
 	// if node is empty
 	if c.Spec.Node.Repository == "" {
-		c.Spec.Node.Repository = config.NodeRepository
-		c.Spec.Node.Tag = config.NodeTag
+		regAndTag := strings.Split(c.GetDefaultImageByPlatformAndName(platform, config.Node), ":")
+		c.Spec.Node.Repository = regAndTag[0]
+		c.Spec.Node.Tag = regAndTag[1]
+
+		changed = true
+	}
+
+	if sidecarChanged := c.clearSidecars(platform); sidecarChanged {
+		changed = sidecarChanged
+	}
+	if sidecarChanged := c.setDefualtSidecars(platform); sidecarChanged {
+		changed = sidecarChanged
+	}
+
+	return changed
+}
+
+// clearPreviousVersion clears the previous version of controller and node
+// images to empty value, so that they will be updated to the latest version.
+// For example: If current controller image is ibmcom/ibm-block-csi-controller-driver:1.0.0,
+// it will be cleared and updated to ibmcom/ibm-block-csi-controller-driver:1.1.0 in setDefault().
+func (c *IBMBlockCSI) clearPreviousVersion(platform string) bool {
+	changed := false
+
+	// if controller is a replace version
+	if config.GetReplaceVersions(platform, config.Controller).Has(c.GetCSIControllerImage()) {
+		c.Spec.Controller.Repository = ""
+		c.Spec.Controller.Tag = ""
+		changed = true
+	}
 
+	// if node is a replace version
+	if config.GetReplaceVersions(platform, config.Node).Has(c.GetCSINodeImage()) {
+		c.Spec.Node.Repository = ""
+		c.Spec.Node.Tag = ""
 		changed = true
 	}
 
 	return changed
 }
+
+func (c *IBMBlockCSI) clearSidecars(platform string) bool {
```

Comment:
`clearSidecarsPreviousVersions`?

Reply:
ok, rename to clearSidecarPreviousVersion

Reply:
I meant sidecars (plural), unless I misunderstood the function